File: OpenBurnBarCore/Sources/OpenBurnBarCore/SharedModels/BudgetRule.swift
import Foundation

// MARK: - BudgetRuleScope

/// Where this budget rule applies. `BudgetGate` evaluates every matching scope when
/// deciding whether to allow / warn / block a request; the most restrictive outcome wins.
public enum BudgetRuleScope: String, Codable, CaseIterable, Hashable, Sendable {
    /// Limit on a specific `(providerID, accountID)` credential. The user's primary lever
    /// for "stop spending on this OpenRouter key at $50/mo."
    case credential
    /// Limit on a specific project name (free-text `token_usage.projectName`).
    case project
    /// Single ceiling across all per-usage credentials. Cheap to add given the per-credential
    /// ledger is the source of truth.
    case global
    /// Org-wide rule synced via Firestore (Phase 8). Enforced locally on each seat against
    /// its own ledger; seat operators cannot disable but can request increases.
    case organization
}

// MARK: - BudgetPeriod

/// Reset cadence for a rule's running spend.
public enum BudgetPeriod: String, Codable, CaseIterable, Hashable, Sendable {
    case day
    case week
    case month
    case allTime

    /// Returns the inclusive lower bound (the start of the current window) for a given
    /// reference time. Used by `BudgetLedger` to slice `token_usage` rows.
    public func windowStart(reference: Date = Date(), calendar: Calendar = .current) -> Date? {
        switch self {
        case .day:
            return calendar.startOfDay(for: reference)
        case .week:
            return calendar.dateInterval(of: .weekOfYear, for: reference)?.start
                ?? calendar.date(byAdding: .day, value: -7, to: reference)
        case .month:
            return calendar.dateInterval(of: .month, for: reference)?.start
                ?? calendar.date(byAdding: .month, value: -1, to: reference)
        case .allTime:
            return nil
        }
    }

    /// Returns the next reset boundary (when the running total resets to zero). `nil` for
    /// `.allTime` since that period never resets.
    public func nextReset(reference: Date = Date(), calendar: Calendar = .current) -> Date? {
        switch self {
        case .day:
            return calendar.date(byAdding: .day, value: 1, to: calendar.startOfDay(for: reference))
        case .week:
            guard let interval = calendar.dateInterval(of: .weekOfYear, for: reference) else { return nil }
            return interval.end
        case .month:
            guard let interval = calendar.dateInterval(of: .month, for: reference) else { return nil }
            return interval.end
        case .allTime:
            return nil
        }
    }
}

// MARK: - BudgetBehavior

/// What `BudgetGate` does when a rule's running spend meets or exceeds its limit.
public enum BudgetBehavior: String, Codable, CaseIterable, Hashable, Sendable {
    /// Emit a one-shot warning at 80% (debounced per-period) and hard-block at 100%.
    /// The user-recommended default per the planning conversation.
    case warnThenBlock
    /// Block immediately at 100% with no warning grace period.
    case hardBlock
    /// Fire the 80% / 100% notifications but always let the request through. Pure observability.
    case warnOnly
    /// Hard-block at 100% AND auto-promote the configured fallback credential(s) on the same
    /// request. Keeps work flowing while still enforcing the cap.
    case hardBlockWithFallback
}

// MARK: - BudgetBillingMode

/// Whether the credential being charged costs marginal $ per token or runs on a flat-rate plan.
/// Subscription credentials short-circuit `BudgetGate.evaluate` to `.allow` so a Claude Pro
/// OAuth key never gets blocked by these limits.
public enum BudgetBillingMode: String, Codable, CaseIterable, Hashable, Sendable {
    case perUsage
    case subscription
    case unknown
}

// MARK: - BudgetRule

/// User-configured spending limit. Persisted in `budget_rules` and synced via `CloudSync`
/// once Phase 8 wires the org-level rollup.
public struct BudgetRule: Identifiable, Codable, Hashable, Sendable {
    public var id: String
    public var scope: BudgetRuleScope
    /// Free-form identifier joined with `providerID`/`accountID`/`projectName` depending on
    /// scope. Kept as a single column so `.organization` rules can carry arbitrary keys
    /// without schema churn.
    public var identifier: String?
    /// For credential-scope rules: the `ProviderID.rawValue` (e.g. "anthropic", "openai",
    /// "openrouter").
    public var providerID: String?
    /// For credential-scope rules: the hashed partition token from `token_usage.providerAccountID`.
    public var accountID: String?
    /// For project-scope rules: the raw `token_usage.projectName` string.
    public var projectName: String?
    /// Human-facing label. Falls back to a synthesized name when nil.
    public var label: String?
    /// Spending ceiling in USD.
    public var amountUSD: Double
    public var period: BudgetPeriod
    public var behavior: BudgetBehavior
    /// Optional ordered list of credential rule IDs to promote when this rule blocks
    /// (only honored when `behavior == .hardBlockWithFallback`).
    public var fallbackCredentialIDs: [String]
    /// When non-nil and in the future, `BudgetGate.evaluate` short-circuits to `.allow` for
    /// this rule. Lets Hermes / Settings temporarily pause enforcement.
    public var pausedUntil: Date?
    public var createdAt: Date
    public var updatedAt: Date
    public var syncedAt: Date?
    public var sourceDeviceID: String?
    public var isEnabled: Bool

    public init(
        id: String = UUID().uuidString,
        scope: BudgetRuleScope,
        identifier: String? = nil,
        providerID: String? = nil,
        accountID: String? = nil,
        projectName: String? = nil,
        label: String? = nil,
        amountUSD: Double,
        period: BudgetPeriod,
        behavior: BudgetBehavior = .warnThenBlock,
        fallbackCredentialIDs: [String] = [],
        pausedUntil: Date? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        syncedAt: Date? = nil,
        sourceDeviceID: String? = nil,
        isEnabled: Bool = true
    ) {
        self.id = id
        self.scope = scope
        self.identifier = identifier
        self.providerID = providerID
        self.accountID = accountID
        self.projectName = projectName
        self.label = label
        self.amountUSD = amountUSD
        self.period = period
        self.behavior = behavior
        self.fallbackCredentialIDs = fallbackCredentialIDs
        self.pausedUntil = pausedUntil
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.syncedAt = syncedAt
        self.sourceDeviceID = sourceDeviceID
        self.isEnabled = isEnabled
    }

    /// Returns a human-facing label. Synthesizes one when `label` is nil so views never
    /// have to guard on it.
    public var displayLabel: String {
        if let label, !label.isEmpty { return label }
        switch scope {
        case .credential:
            let providerName = providerID ?? "credential"
            if let accountID, !accountID.isEmpty {
                return "\(providerName) · …\(accountID.suffix(6))"
            }
            return "\(providerName) · default"
        case .project:
            return projectName ?? "Unnamed project"
        case .global:
            return "All per-usage credentials"
        case .organization:
            return identifier ?? "Organization"
        }
    }

    /// True when `pausedUntil` is set and still in the future.
    public func isPaused(at reference: Date = Date()) -> Bool {
        if let pausedUntil { return pausedUntil > reference }
        return false
    }
}

// MARK: - BudgetEventKind

/// Append-only audit log entry kinds. Feeds `burnbar_budget_audit` and the in-app event
/// inspector.
public enum BudgetEventKind: String, Codable, CaseIterable, Hashable, Sendable {
    case warning
    case block
    case override
    case pause
    case resume
    case ruleCreated
    case ruleUpdated
    case ruleDeleted
}

// MARK: - BudgetEvent

/// Single immutable audit-log row.
public struct BudgetEvent: Identifiable, Codable, Hashable, Sendable {
    public var id: String
    public var ruleID: String
    public var kind: BudgetEventKind
    /// Where this event originated — "settings_ui", "hermes_tool", "mcp_tool", "auto_gate",
    /// "billing_reconciliation". Free-form so future sources don't require schema changes.
    public var source: String?
    public var amountAtEvent: Double
    public var limitAtEvent: Double
    public var detailJSON: String?
    public var occurredAt: Date
    public var syncedAt: Date?
    public var sourceDeviceID: String?

    public init(
        id: String = UUID().uuidString,
        ruleID: String,
        kind: BudgetEventKind,
        source: String? = nil,
        amountAtEvent: Double,
        limitAtEvent: Double,
        detailJSON: String? = nil,
        occurredAt: Date = Date(),
        syncedAt: Date? = nil,
        sourceDeviceID: String? = nil
    ) {
        self.id = id
        self.ruleID = ruleID
        self.kind = kind
        self.source = source
        self.amountAtEvent = amountAtEvent
        self.limitAtEvent = limitAtEvent
        self.detailJSON = detailJSON
        self.occurredAt = occurredAt
        self.syncedAt = syncedAt
        self.sourceDeviceID = sourceDeviceID
    }
}

// MARK: - BudgetCredentialIdentity

/// Unified credential handle used by `BudgetGate` to look up matching rules. Bridges
/// the daemon plane (`BurnBarProviderRoute.credentialSlotID` + `providerID`) and the
/// AgentLens plane (raw bearer token → hashed slot ID).
public struct BudgetCredentialIdentity: Hashable, Sendable {
    public let providerID: String
    public let slotID: String
    public let displayLabel: String
    public let billingMode: BudgetBillingMode

    public init(
        providerID: String,
        slotID: String,
        displayLabel: String,
        billingMode: BudgetBillingMode = .unknown
    ) {
        self.providerID = providerID
        self.slotID = slotID
        self.displayLabel = displayLabel
        self.billingMode = billingMode
    }

    /// Derives `BudgetBillingMode` from a raw secret prefix. Anthropic OAuth keys
    /// (`sk-ant-oat*`) are subscription identities and must never be gated; Console keys
    /// (`sk-ant-api*`) and bearer-style OpenAI keys (`sk-*`) are per-usage.
    public static func billingMode(forSecretPrefix prefix: String) -> BudgetBillingMode {
        let lower = prefix.lowercased()
        if lower.hasPrefix("sk-ant-oat") { return .subscription }
        if lower.hasPrefix("sk-ant-api") { return .perUsage }
        if lower.hasPrefix("sk-") { return .perUsage }
        return .unknown
    }
}

// MARK: - BudgetGateDecision

/// `BudgetGate.evaluate` outcome. Pure data — callers translate into 402 / chat error /
/// header attachment.
public enum BudgetGateDecision: Hashable, Sendable {
    /// Default — request flows through with no gate intervention.
    case allow
    /// Pass-through, but a header / chip should be attached so the chat surface can render
    /// "you're at \(usedPercent)% of \(rule.displayLabel)."
    case warn(rule: BudgetRule, usedPercent: Double, used: Double, limit: Double)
    /// Hard block. Caller should return 402 (daemon) or throw `BudgetBlockedError` (AgentLens).
    /// `fallback` is the next candidate credential when `rule.behavior == .hardBlockWithFallback`.
    case block(rule: BudgetRule, used: Double, limit: Double, fallback: BudgetCredentialIdentity?)
    /// Rule is enforcement-eligible but currently paused (e.g. Hermes set a 1-hour bypass).
    /// Treated as allow but surfaced separately for telemetry.
    case paused(rule: BudgetRule, resumeAt: Date)
}

File: windows/app/OpenBurnBar.App.Presentation/Budget/BudgetRule.cs
using System;

namespace OpenBurnBar.App.Presentation.Budget;

// PORTED (faithful, scoped) from OpenBurnBarCore/.../SharedModels/BudgetRule.swift.
// The enums, the BudgetRule record, and the period window/reset math are the pure,
// storage-agnostic half of the Budget surface. The SAME managed assembly is referenced
// by the WinUI app (whose XAML x:Binds to the view-models built on these types) AND is
// unit-tested on the macOS authoring host via `dotnet test` (windows/tests/presentation).
//
// The enum RAW-VALUE strings are kept byte-identical to the Swift `String`-backed enums
// so a future SQLCipher `budget_rules` table stays cross-platform readable (the storage
// seam lands in a later wave); BudgetEnumRaw round-trips them.

/// <summary>Where a budget rule applies. Mirrors Swift <c>BudgetRuleScope</c>.</summary>
public enum BudgetRuleScope
{
    /// <summary>Limit on a specific <c>(providerID, accountID)</c> credential.</summary>
    Credential,

    /// <summary>Limit on a specific free-text project name.</summary>
    Project,

    /// <summary>Single ceiling across every per-usage credential.</summary>
    Global,

    /// <summary>Org-wide rule synced via the cloud (enforced locally per seat).</summary>
    Organization,
}

/// <summary>Reset cadence for a rule's running spend. Mirrors Swift <c>BudgetPeriod</c>.</summary>
public enum BudgetPeriod
{
    Day,
    Week,
    Month,
    AllTime,
}

/// <summary>What the gate does when a rule's running spend meets/exceeds its limit.</summary>
public enum BudgetBehavior
{
    /// <summary>Warn at the threshold (default 80%), hard-block at 100%.</summary>
    WarnThenBlock,

    /// <summary>Block immediately at 100% with no warning grace period.</summary>
    HardBlock,

    /// <summary>Fire threshold/limit notifications but always let the request through.</summary>
    WarnOnly,

    /// <summary>Hard-block at 100% AND promote the configured fallback credential(s).</summary>
    HardBlockWithFallback,
}

/// <summary>Whether the charged credential costs marginal $ or runs on a flat-rate plan.</summary>
public enum BudgetBillingMode
{
    PerUsage,
    Subscription,
    Unknown,
}

/// <summary>
/// Byte-for-byte parity with the Swift <c>rawValue</c> strings so the persisted
/// <c>budget_rules</c> / <c>budget_audit</c> rows stay cross-platform readable.
/// </summary>
public static class BudgetEnumRaw
{
    public static string Raw(this BudgetRuleScope scope) => scope switch
    {
        BudgetRuleScope.Credential => "credential",
        BudgetRuleScope.Project => "project",
        BudgetRuleScope.Global => "global",
        BudgetRuleScope.Organization => "organization",
        _ => "global",
    };

    public static string Raw(this BudgetPeriod period) => period switch
    {
        BudgetPeriod.Day => "day",
        BudgetPeriod.Week => "week",
        BudgetPeriod.Month => "month",
        BudgetPeriod.AllTime => "allTime",
        _ => "month",
    };

    public static string Raw(this BudgetBehavior behavior) => behavior switch
    {
        BudgetBehavior.WarnThenBlock => "warnThenBlock",
        BudgetBehavior.HardBlock => "hardBlock",
        BudgetBehavior.WarnOnly => "warnOnly",
        BudgetBehavior.HardBlockWithFallback => "hardBlockWithFallback",
        _ => "warnThenBlock",
    };

    public static string Raw(this BudgetBillingMode mode) => mode switch
    {
        BudgetBillingMode.PerUsage => "perUsage",
        BudgetBillingMode.Subscription => "subscription",
        BudgetBillingMode.Unknown => "unknown",
        _ => "unknown",
    };

    public static BudgetRuleScope ParseScope(string raw) => raw switch
    {
        "credential" => BudgetRuleScope.Credential,
        "project" => BudgetRuleScope.Project,
        "global" => BudgetRuleScope.Global,
        "organization" => BudgetRuleScope.Organization,
        _ => BudgetRuleScope.Global,
    };

    public static BudgetPeriod ParsePeriod(string raw) => raw switch
    {
        "day" => BudgetPeriod.Day,
        "week" => BudgetPeriod.Week,
        "month" => BudgetPeriod.Month,
        "allTime" => BudgetPeriod.AllTime,
        _ => BudgetPeriod.Month,
    };

    public static BudgetBehavior ParseBehavior(string raw) => raw switch
    {
        "warnThenBlock" => BudgetBehavior.WarnThenBlock,
        "hardBlock" => BudgetBehavior.HardBlock,
        "warnOnly" => BudgetBehavior.WarnOnly,
        "hardBlockWithFallback" => BudgetBehavior.HardBlockWithFallback,
        _ => BudgetBehavior.WarnThenBlock,
    };
}

/// <summary>
/// Window/reset arithmetic for a <see cref="BudgetPeriod"/>. Mirrors the Swift
/// <c>BudgetPeriod.windowStart</c> / <c>nextReset</c> that <c>BudgetLedger</c> uses to slice
/// <c>token_usage</c>. Deterministic (no ambient <c>Calendar.current</c>): the reference's own
/// UTC offset is preserved and the week start day is explicit (default Sunday, matching the
/// US-locale Gregorian calendar the Swift path resolves to).
/// </summary>
public sealed class BudgetClock
{
    private readonly DayOfWeek _firstDayOfWeek;

    public BudgetClock(DayOfWeek firstDayOfWeek = DayOfWeek.Sunday)
    {
        _firstDayOfWeek = firstDayOfWeek;
    }

    public static BudgetClock Default { get; } = new();

    /// <summary>Inclusive lower bound (start of the current window). <c>null</c> for all-time.</summary>
    public DateTimeOffset? WindowStart(BudgetPeriod period, DateTimeOffset reference)
    {
        switch (period)
        {
            case BudgetPeriod.Day:
                return StartOfDay(reference);
            case BudgetPeriod.Week:
                return StartOfWeek(reference);
            case BudgetPeriod.Month:
                return StartOfMonth(reference);
            case BudgetPeriod.AllTime:
            default:
                return null;
        }
    }

    /// <summary>Next reset boundary (when the running total resets). <c>null</c> for all-time.</summary>
    public DateTimeOffset? NextReset(BudgetPeriod period, DateTimeOffset reference)
    {
        switch (period)
        {
            case BudgetPeriod.Day:
                return StartOfDay(reference).AddDays(1);
            case BudgetPeriod.Week:
                return StartOfWeek(reference).AddDays(7);
            case BudgetPeriod.Month:
                return StartOfMonth(reference).AddMonths(1);
            case BudgetPeriod.AllTime:
            default:
                return null;
        }
    }

    private static DateTimeOffset StartOfDay(DateTimeOffset reference) =>
        new(reference.Year, reference.Month, reference.Day, 0, 0, 0, reference.Offset);

    private DateTimeOffset StartOfWeek(DateTimeOffset reference)
    {
        DateTimeOffset day = StartOfDay(reference);
        int delta = ((int)day.DayOfWeek - (int)_firstDayOfWeek + 7) % 7;
        return day.AddDays(-delta);
    }

    private static DateTimeOffset StartOfMonth(DateTimeOffset reference) =>
        new(reference.Year, reference.Month, 1, 0, 0, 0, reference.Offset);
}

/// <summary>
/// User-configured spending limit. Faithful port of the Swift <c>BudgetRule</c> value type
/// (a <c>record</c> here so the WinUI editor mutates a copy with <c>with { }</c> and tests get
/// value equality). Persisted in <c>budget_rules</c> in a later storage wave.
/// </summary>
public sealed record BudgetRule
{
    public string Id { get; init; } = Guid.NewGuid().ToString();
    public BudgetRuleScope Scope { get; init; }
    public string? Identifier { get; init; }
    public string? ProviderId { get; init; }
    public string? AccountId { get; init; }
    public string? ProjectName { get; init; }
    public string? Label { get; init; }

    /// <summary>Spending ceiling in USD.</summary>
    public double AmountUsd { get; init; }

    public BudgetPeriod Period { get; init; } = BudgetPeriod.Month;
    public BudgetBehavior Behavior { get; init; } = BudgetBehavior.WarnThenBlock;

    /// <summary>Ordered credential rule IDs to promote when this rule blocks (fallback only).</summary>
    public System.Collections.Generic.IReadOnlyList<string> FallbackCredentialIds { get; init; } =
        Array.Empty<string>();

    /// <summary>When non-null and in the future, the gate short-circuits this rule to allow.</summary>
    public DateTimeOffset? PausedUntil { get; init; }

    public DateTimeOffset CreatedAt { get; init; } = DateTimeOffset.UtcNow;
    public DateTimeOffset UpdatedAt { get; init; } = DateTimeOffset.UtcNow;
    public DateTimeOffset? SyncedAt { get; init; }
    public string? SourceDeviceId { get; init; }
    public bool IsEnabled { get; init; } = true;

    /// <summary>Human-facing label; synthesizes one when <see cref="Label"/> is empty.</summary>
    public string DisplayLabel
    {
        get
        {
            if (!string.IsNullOrEmpty(Label))
            {
                return Label!;
            }

            switch (Scope)
            {
                case BudgetRuleScope.Credential:
                    string providerName = string.IsNullOrEmpty(ProviderId) ? "credential" : ProviderId!;
                    if (!string.IsNullOrEmpty(AccountId))
                    {
                        string tail = AccountId!.Length <= 6 ? AccountId! : AccountId!.Substring(AccountId!.Length - 6);
                        return $"{providerName} · …{tail}";
                    }

                    return $"{providerName} · default";
                case BudgetRuleScope.Project:
                    return string.IsNullOrEmpty(ProjectName) ? "Unnamed project" : ProjectName!;
                case BudgetRuleScope.Global:
                    return "All per-usage credentials";
                case BudgetRuleScope.Organization:
                    return string.IsNullOrEmpty(Identifier) ? "Organization" : Identifier!;
                default:
                    return "Budget rule";
            }
        }
    }

    /// <summary>True when <see cref="PausedUntil"/> is set and still in the future.</summary>
    public bool IsPaused(DateTimeOffset reference) => PausedUntil is DateTimeOffset until && until > reference;
}
